### PR TITLE
Add parameter gradient test to legacy driver

### PR DESCRIPTION
**Pull request by markdewing**
Add the gradient test (#3540 ) to the legacy linear optimizer.
When testing the optimizer and parameter gradients, it's helpful to be able to compare to the legacy driver to help isolate where problems occur.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- N/A. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Not yet. Documentation has been added (if appropriate)

### CODE REVIEW

**Review by ye-luo — CHANGES_REQUESTED** [review]
Please add documentation as #3793 requested.

**Comment by markdewing** [comment]
I'd like #3917 to get in before I write more documentation.

**Comment by prckent** [comment]
Happy to approve this once the conflict is fixed.

I assume one of the additional next steps will be to activate this code.

**Review by prckent — APPROVED** [review]
Note Ye's comment to add docs at some point.

**Comment by ye-luo** [comment]
Please extend the existing test to cover the classic driver.

**Comment by prckent** [comment]
Ping

**Comment by markdewing** [comment]
Test this please

File: docs/methods.rst [hunk 1]
```diff
@@ -1206,6 +1206,43 @@ the values found in the \*.sXXX.opt.h5 file. Be careful to keep the pair
...
+The gradients of the energy with respect to the variational parameters can be checked and optionally written to a file.
+The check compares the analytic derivatives with a finite difference approximation.
+These are activated by giving a ``gradient_test`` method in and ``optimize`` block, as follows:
+
+::
+
+     <qmc method="linear" move="pbyp">
+      <optimize method="gradient_test">
+      </optimize>
+      ... rest of optimizer input ...
```
Comment (ye-luo) [inline docs/methods.rst:1220]: `rest of optimizer input`->`VMC related input`?
With gradient_test, I doubt any other optimizer parameters are still needed.

File: tests/molecules/He_param/CMakeLists.txt [hunk 2]
```diff
@@ -25,6 +25,25 @@ if(NOT QMC_CUDA)
...
+  # Check the parameter gradient code for the legacy driver
+  qmc_run_and_check_custom_scalar(
+    BASE_NAME
+    He_param_grad_legacy
+    BASE_DIR
+    "${qmcpack_SOURCE_DIR}/tests/molecules/He_param"
+    PREFIX
+    He_param_grad.param
+    INPUT_FILE
+    He_param_grad.xml
```
Comment (ye-luo) [inline tests/molecules/He_param/CMakeLists.txt:37]: Incorrect input file. This explains the coverage report complaints.
Can we rename the input file He_param_grad_legacy_driver.xml

**Comment by markdewing** [comment]
I'm not sure it's worth the effort to extend the deterministic parameter gradient test to the legacy driver.

File: docs/methods.rst (hunk 1 above)
Comment (markdewing) [reply docs/methods.rst:1220]: This is probably true, but I expect the common case is starting from an existing optimizer input block.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Comment by markdewing** [comment]
My original motivation for adding this to the legacy driver is to compare wavefunction and gradient values from the batched driver and the legacy driver to see where they differ. A deterministic test is a regression test for a particular wavefunction - that sort of testing seems useful for ensuring changes to the driver don't change the code unexpectedly.  And not as useful for investigating existing wavefunction code.

**Review by prckent — COMMENTED** [review]
When does this test get run?

It looks to be a "short" test and not a "deterministic" test. Specific prefixes need to be used due to using things like -R short or -E long in the various nightly scripts.

How long does it take? What are the expected statistics?

**Comment by markdewing** [comment]
The test that was added is a statistical test, not a deterministic test.   It is a copy of the existing statistical test for the parameter gradients for using the batched driver (in molecules/He_param). The values of the gradients should be the same.

There is a deterministic parameter gradient test using the batched driver for a different system (solids/diamondC_1x1x1_pp).  Copying and modifying that test for the legacy driver is what I pushed back on.

**Comment by ye-luo** [comment]
Indeed, `tests/solids/diamondC_1x1x1_pp/det_qmc_param_grad.xml` was the deterministic version.
He_param was actually a bit off in the nightly. https://cdash.qmcpack.org/CDash/queryTests.php?project=QMCPACK&date=2022-05-06&filtercount=1&showfilters=1&field1=testname&compare1=63&value1=param_grad
I consider the deterministic one is more valuable than the statistically one. The statistically one is impossible to run with offload to the very slow finite difference.

**Comment by markdewing** [comment]
I prefer PR's to be focused.  In this case the added test is to just enough to test the additions in this PR.  Since the added functionality is for testing, this makes the difference between functionality and tests a bit more subtle.  In my view, deterministic tests use this PR to test other functionality, but are not as good at testing this PR.   
Adding the deterministic test should be a separate PR.

**Comment by ye-luo** [comment]
We may or or may not use this feature to create deterministic tests. I saw that a orthogonal topic.

This "gradient_test" is a QMCPACK feature rather than a test managed by catch. In my view, It is better to be protected by a deterministic test before being used for testing other things.

Right now the feature has been protected by a statistical test although inferior to a deterministic test. So this PR can get in but I will leave others to decide if a deterministic test is a strong demand in this PR or not.

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
A deterministic test has been added.

**Pull request merged**